### PR TITLE
Feature/forward composer auth

### PR DESCRIPTION
**Pull request by ikappas**
| Q                 | A
| ----------------- | ---
| Issue Type        | Feature Request
| Deployer Version  | N/A
| Local Machine OS  | N/A
| Remote Machine OS | N/A

### Short Description 
Add an option to forward composer auth (similar to forward agent) to the remote servers in order to access private repositories.

### Description
Using composer private repositories via Satis or Toran Proxy does not work unless `forward agent` is used. However, these repositories are cloned instead of downloaded which is slow.
```bash
[staging] <   - Installing example/package (1.0)
[staging] <     Downloading
[staging] <     Failed to download example/package from dist: The 'https://packages.example.com/packages/composer/dist/archive.zip?project_id=212&ref=1.0' URL required authentication.
[staging] < You must be using the interactive console to authenticate
[staging] <     Now trying to download from source
[staging] <   - Installing example/package (1.0)
[staging] <     Cloning 738e67d11e33b924bac17eddcccc2a0dfaef3bd
```
This can be mitigated by forwarding composer auth to the deployment server using COMPOSER_AUTH environment variable.
### References
- https://getcomposer.org/doc/articles/http-basic-authentication.md
- https://getcomposer.org/doc/03-cli.md#composer-auth

### CODE REVIEW

**Comment by elfet** [comment]
Please, fix failed tests.

**Comment by ikappas** [comment]
@elfet can you have a  look why the test fail in ci?

**Comment by elfet** [comment]
```
Fatal error: Uncaught exception 'RuntimeException' with message 'The HOME or COMPOSER_HOME environment variable must be set for composer to run correctly' in /tmp/php-wrap-XS7tAh:701
```

**Comment by ikappas** [comment]
Hmm thats weird cause the HOME environment variable should be set.

**Comment by elfet** [comment]
How to test in locally? Or on remote server? What to install?

**Comment by ikappas** [comment]
Ok... here goes... The tests fail because 
```
#0 /tmp/php-wrap-JVRCnF(670): getUserDir()
#1 /tmp/php-wrap-JVRCnF(454): getHomeDir()
#2 /tmp/php-wrap-JVRCnF(55): installComposer()
#3 /tmp/php-wrap-JVRCnF(13): process()
#4 {main}
```
which is triggered by the command:
```bash
cd /home/travis/build/deployphp/deployer/test/localhost && (cd /home/travis/build/deployphp/deployer/test/localhost/releases/2 && curl -sS https://getcomposer.org/installer | /usr/bin/php)
```
So this exception comes from composer installer 😉 
See https://github.com/composer/composer/blame/master/src/Composer/Factory.php#L641

Also note that the **current** way to install composer has changed and is documented:
https://getcomposer.org/doc/faqs/how-to-install-composer-programmatically.md

File: src/Server/Local.php [hunk 1]
```diff
@@ -50,19 +50,20 @@ public function connect()
     /**
      * {@inheritdoc}
      */
-    public function run($command)
+    public function run($command, array $env = [])
```
Comment (elfet) [inline src/Server/Local.php]: I don't like adding `$env` to run function. Why not just pass env as `ENV=val composer ...`?

**Comment by ikappas** [comment]
@elfet Quick question... any reason why you use --prefer-source instead of --prefer-dist with composer update in your tests?

**Comment by elfet** [comment]
This is a big feature, and i'm really wanting to be in in v5. But need a lot of refactoring, ...so i'm going to implement it myself based on this PR.

**Comment by elfet** [comment]
I think better wait for v5 before continuing this this feature.

**Comment by antonmedv** [comment]
Is there anyone who wants to implement it?

**Comment by antonmedv** [comment]
Closing it as this PR is outdated.

**Pull request closed without merging**